kura/org.eclipse.kura.driver.opcua.provider/src/main/java/org/eclipse/kura/internal/driver/opcua/OpcUaDriver.java

```diff
@@ -468,7 +477,52 @@ private static void fail(final DriverRecord record, final String message) {
                 return Optional.empty();
             }
 
-            return Optional.of(new OpcUaRequestInfo(dataType, nodeNamespaceIndex, nodeId));
+            return Optional.of(new OpcUaRequestInfo(record, dataType, nodeNamespaceIndex, nodeId));
+        }
+    }
+
+    @Override
+    public PreparedRead prepareRead(List<DriverRecord> driverRecords) {
+        requireNonNull(driverRecords);
+
+        OpcUaPreparedRead preparedRead = new OpcUaPreparedRead();
+        preparedRead.driverRecords = driverRecords;
+
+        for (DriverRecord record : driverRecords) {
+            OpcUaRequestInfo.extract(record).ifPresent(preparedRead.requestInfos::add);
+        }
+        return preparedRead;
+    }
+
+    private class OpcUaPreparedRead implements PreparedRead {
+
+        private List<OpcUaRequestInfo> requestInfos = new ArrayList<OpcUaRequestInfo>();
+        private List<DriverRecord> driverRecords;
```

Comment:
I think this should be volatile. If there is only one thread acting in `execute()` method and another thread invokes `getDriverRecords()`, there can be situation when the `getDriverRecords()` calling thread doesn't get the updated `DriverRecord`s.